source/testing/testing-components.md

```diff
@@ -180,3 +180,96 @@ test('should trigger external action on form submit', function(assert) {
+  locationService: Ember.inject.service('location-service'),
+
+  //when the coordinates change, call the location service to evaluate what the city and country would be
+  city: Ember.computed('locationService.currentLocation', function () {
+    return this.get('locationService').getCurrentCity();
+  }),
+
+  country: Ember.computed('locationService.currentLocation', function () {
+    return this.get('locationService').getCurrentCountry();
+  })
+});
+```
+
+```app/templates/components/location-indicator.hbs
+You currently are located in {{city}}, {{country}}
+```
+To stub the location service in your test, create a local stub object that extends
+`Ember.Service`, and register the stub as the service your tests need in the
+beforeEach function.  In this case we initially force location to New York.
+
+
+```tests/integration/components/location-indicator-test.js
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import Ember from 'ember';
+
+//Stub location service
+let locationStub = Ember.Service.extend({
+  city: 'New York',
+  country: 'USA',
+  currentLocation: {
+    x: 1234,
+    y: 5678
+  }
+  getCurrentCity() {
+    return this.get('city');
+  },
+  getCurrentCountry() {
+    return this.get('country');
+  }
+});
+
+moduleForComponent('location-indicator', 'Integration | Component | location indicator', {
+  integration: true,
+
+  beforeEach: function () {
+    this.registry.register('service:location-service', locationStub);
+  }
+});
+```
+
+Once the stub service is registered the test simply needs to check that the stub data that
+is being returned from the service is reflected in the component output.
+
+```tests/integration/components/location-indicator-test.js
+test('should reveal current location', function(assert) {
+  this.render(hbs`{{location-indicator}}`);
+  assert.equal(this.$().text().trim(), 'You currently are located in New York, USA');
+});
+```
+
+Since services default to being singletons, we can get a reference to the service within our test
+and modify values on it so that we can assert on how we expect our component to react. In the next
+example, we'll add another test that validates that the display changes when we modify the values
+on the service.
+
+```tests/integration/components/location-indicator-test.js
+test('should change displayed location when current location changes', function (assert) {
+  this.render(hbs`{{location-indicator}}`);
+  assert.equal(this.$().text().trim(), 'You currently are located in New York, USA', 'origin location should display');
+  let locationService = this.container.lookup('service:location-service');
+  Ember.run(() => {
+    locationService.set('city', 'Beijing');
+    locationService.set('country', 'China');
+    locationService.set('currentLocation', { x: 11111, y: 222222 });
+  });
+  assert.equal(this.$().text().trim(), 'You currently are located in Beijing, China', 'location display should change');
```

Comment:
So as a general practice, I add the end arguments on asserts to explain what they are asserting.  Otherwise they show up in the testem ui output as just 'ok'.  When they fail (or pass), its nice to have the extra info.  In this case there's 2 asserts, first one asserts that the location displays after render, second asserts that the location is updated onscreen after the service changes.